File: CP/Codeforces/166E.cpp
#include <bits/stdc++.h>

using namespace std;

#define odd(a) ((a) & 1)
#define sqr(a) (a) * (a)

typedef long long ll;

struct TMatrix
{
    ll a[4][4];
    TMatrix() { memset(a, 0, sizeof a); }
    ll* operator [](int i) { return a[i]; }
};

const ll mod = 1E9 + 7;

int n;
TMatrix ans;

TMatrix operator *(TMatrix &a, TMatrix &b)
{
    TMatrix c;
    for (int i = 0; i < 4; ++i)
        for (int j = 0; j < 4; ++j)
            for (int k = 0; k < 4; ++k)
                (c[i][j] += a[i][k] * b[k][j] % mod) %= mod;
    return c;
}

TMatrix power(TMatrix a, int b)
{
    if (b == 1) return a;
    TMatrix tmp = power(a, b >> 1), ret = sqr(tmp);
    if (odd(b)) ret = ret * a;
    return ret;
}

void solve()
{
    for (int i = 0; i < 4; ++i)
        for (int j = 0; j < 4; ++j) ans[i][j] = i == j ? 0 : 1;
    ans = power(ans, n);
}

int main()
{
    ios_base::sync_with_stdio(false);
    scanf("%d", &n);
    solve();
    printf("%lld\n", ans[3][3]);
    return 0;
}

File: CP/Codeforces/166E.pas
{$MODE OBJFPC}
type
  TMatrix=array[1..4,1..4] of int64;
const
  module=trunc(1E9)+7;
  base:TMatrix=((0,1,1,1),(1,0,1,1),(1,1,0,1),(1,1,1,0));
var
  n:longint;
  ans:TMatrix;

operator *(a,b:TMatrix) c:TMatrix;
var
  i,j,k:longint;
begin
  fillchar(c,sizeof(c),0);
  for i:=1 to 4 do
    for j:=1 to 4 do
      for k:=1 to 4 do
        c[i,j]:=(c[i,j]+a[i,k]*b[k,j] mod module) mod module;
end;

function pow(a:TMatrix;b:longint):TMatrix;
var
  t:TMatrix;
begin
  if b=1 then exit(a);
  t:=pow(a,b div 2);
  t:=t*t;
  if odd(b) then t:=t*a;
  pow:=t;
end;

begin
  readln(n);
  ans:=pow(base,n);
  writeln(ans[4,4]);
end.
